### PR TITLE
Callback documentation

### PR DESCRIPTION
**Pull request by blisc**
To-do:
- [x] A v0.10 to v0.11 upgrade guide

### CODE REVIEW

File: docs/sources/source/tutorials/callbacks.rst [hunk 1]
```diff
@@ -1,171 +1,182 @@
...
+
+                ...  # Forward and backward pass
+
+                callbacks.on_batch_end()
+                batch_counter += 1
+                if batch_counter == gradient_accumulation_steps:
+                    # By default, gradient_accumulation_steps = 1
+                    # Note this is passed to train() as batches_per_step and sometimes exposed as args.iter_per_step
+                    optimizer.step()
+                    callbacks.update_loss()  # Note, that the loss tensor is only exposed here
```
Comment (tkornuta-nvidia) [inline docs/sources/source/tutorials/callbacks.rst:40]: Hey, @blisc what is that "update_loss()" doing? Is that a special _event type_?

And shouldn't it be "loss tensors are"?

File: docs/sources/source/tutorials/old_callbacks.rst [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+We recommend for user_epochs_done_callback to simply return the global_var_dict
+for tb_writer_func to consume. The user must log all data of interest inside
+tb_writer_func including scalars that would otherwise be logged if
+tb_writer_func was not passed to EvaluatorCallback.
+
+You can also log your evaluation metrics into Weights & Biases experiment trackers.
+To do so, please setup these parameters. Also make sure wandb is installed and you did ``wandb login``.
+
+- wandb_name: W&B experiment name
+- wandb_project: W&B project name
```
Comment (tkornuta-nvidia) [inline docs/sources/source/tutorials/old_callbacks.rst:157]: Shouldn't that go into the next section about Wandb?

File: docs/sources/source/tutorials/callbacks.rst (hunk 1 above)
Comment (blisc) [reply docs/sources/source/tutorials/callbacks.rst:40]: There is only 1 loss tensor, which is the summation of `tensors_to_optimize` that is passed to train().
update_loss() just does the equivalent of `state["tensors"]["loss"] = final_loss`. It just updates the loss tensor in the tensor dictionary.

File: docs/sources/source/tutorials/old_callbacks.rst [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+
+    def log_to_tb_func(swriter, tensors, step):
+        swriter.add_scalar("Train_Loss", tensors[0], step)
+        swriter.add_audio("Train_Sample", tensors[1][0], step)
+
+SimpleLossLoggerCallback can be constructed as follows:
+
+.. code-block:: python
+
+    from functools import partial
```
Comment (tkornuta-nvidia) [inline docs/sources/source/tutorials/old_callbacks.rst:75]: Do we need this import here?

File: docs/sources/source/tutorials/old_callbacks.rst [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+        print_func=my_print_func,
+        # Pass the function that returns tensorboard tags and scalars
+        get_tb_values=my_get_tb_values,
+        # How often we want to call this callback
+        step_freq=500,
+        # The tensorboard writer object we want to use, it should be
+        # automatically created by neural_factory if create_tb_writer was
+        # set to True during neural_factory construction
+        tb_writer=neural_factory.tb_writer)
+    )
```
Comment (tkornuta-nvidia) [inline docs/sources/source/tutorials/old_callbacks.rst:90]: This is great, I remember was impressed by callback description in PyTorchLightning, nice!🥇

File: nemo/core/callbacks.py [hunk 5]
```diff
@@ -65,7 +65,7 @@ class NeMoCallback(ABC):
     TODO: Add a link to documentation.
     """
 
-    def on_train_start(self, state):
+    def on_action_start(self, state):
```
Comment (tkornuta-nvidia) [inline nemo/core/callbacks.py:68]: 👍 👍 👍 👍 👍

File: docs/sources/source/tutorials/callbacks.rst (hunk 1 above)
Comment (tkornuta-nvidia) [reply docs/sources/source/tutorials/callbacks.rst:40]: Wait... this relates to my questions from some time ago... So can you pass more loss tensors to `tensors_to_optimize` or not...? Let's take that discussion offline.

In here more important question is: is `on_update_loss` an event type? If so, then it should be added to that nice list of events that you provided...

File: docs/sources/source/tutorials/old_callbacks.rst (hunk 2 above)
Comment (blisc) [reply docs/sources/source/tutorials/old_callbacks.rst:157]: Are these changes (1/3) important? Note that this is old_callbacks.rst, which covers the deprecated callbacks and are planned for removal in v0.12.

File: docs/sources/source/tutorials/old_callbacks.rst (hunk 3 above)
Comment (blisc) [reply docs/sources/source/tutorials/old_callbacks.rst:75]: Are these changes (2/3) important? Note that this is old_callbacks.rst, which covers the deprecated callbacks and are planned for removal in v0.12.

I think partial is a cool function that most people don't know exists (incl. me before I found out) and it is used in this codeblock.

File: docs/sources/source/tutorials/callbacks.rst (hunk 1 above)
Comment (blisc) [reply docs/sources/source/tutorials/callbacks.rst:40]: Yep you can pass as many tensors as you want to `tensors_to_optimize`. But the loss that we call backwards on is the summation of all tensors in `tensors_to_optimize`.

At a very simplified overview it more or less works like this, assuming no gradient accumulation:
```
final_loss = 0
for t in tensors_to_optimize:
    final_loss += t
final_loss.backwards()
optimizer.step()
state["tensors"]["loss"] = final_loss
```

I don't think `on_update_loss` is important as it occurs at exactly the same time as `on_step_end` so it seems redundant.

File: docs/sources/source/tutorials/old_callbacks.rst (hunk 4 above)
Comment (tkornuta-nvidia) [reply docs/sources/source/tutorials/old_callbacks.rst:90]: wait... I didn't get that, I thought this is the new one... my bad :]

File: docs/sources/source/tutorials/old_callbacks.rst (hunk 3 above)
Comment (tkornuta-nvidia) [reply docs/sources/source/tutorials/old_callbacks.rst:75]: Yap, a cool function, but I cannot see _any reference_ to this function...

File: docs/sources/source/tutorials/callbacks.rst [hunk 6]
```diff
@@ -1,171 +1,182 @@
...
+                    callbacks.on_step_end()
+                clear_tensor_dict()
```
Comment (blisc) [inline docs/sources/source/tutorials/callbacks.rst:42]: Add comments about the tensor dict.

File: docs/sources/source/tutorials/callbacks.rst [hunk 7]
```diff
@@ -1,171 +1,182 @@
...
+At a high level, the NeMo training loop looks like this:
 
 .. code-block:: python
 
-    from nemo import logging
-    def my_print_func(tensors):
-        logging.info(f"Loss {tensors[0]}")
+    def train():
+        # Do initialization of optimizers, amp, ddp, etc
+        callbacks.on_action_start()  # Do all on_action_start functions
```
Comment (blisc) [inline docs/sources/source/tutorials/callbacks.rst:23]: Make the comment explicit about events

File: docs/sources/source/tutorials/old_callbacks.rst (hunk 3 above)
Comment (blisc) [reply docs/sources/source/tutorials/old_callbacks.rst]: it has been removed

File: docs/sources/source/tutorials/callbacks.rst [hunk 8]
```diff
@@ -1,171 +1,190 @@
...
+                # Set the `loss` key inside the tensor state object to be the loss that we call backwards() on
+                state["tensors"]["loss"] = final_loss
+
+                callbacks.on_batch_end(state)  # Trigger the on_batch_end event
+                batch_counter += 1
+                if batch_counter == gradient_accumulation_steps:
+                    # By default, gradient_accumulation_steps = 1. Note this is passed to train() as batches_per_step
+                    # and sometimes exposed as args.iter_per_step
+                    optimizer.step()
+                    callbacks.on_step_end(state)  # Trigger the on_step_end event
```
Comment (tkornuta-nvidia) [inline docs/sources/source/tutorials/callbacks.rst:47]: Just to make it perfectly clear I would add here:
```batch_counter = 0```
